### ISSUE TITLE
Show/Hide columns malfunction when using useState with enableGrouping

### ISSUE THREAD

**Issue by realDNA** [KevinVandy/material-react-table#217]
### material-react-table version

1.3.10

### react & react-dom versions

18.2.0

### Describe the bug and the steps to reproduce it

```
import React, { useEffect, useState } from "react";

import MaterialReactTable from "material-react-table";

const Basic = () => {
  const [columns, setColumns] = useState([]);
  const [data, setData] = useState([]);

  useEffect(() => {

     // Do something and set columns and data

    setColumns([
      {
        accessorKey: "name.firstName", //access nested data with dot notation
        header: "First Name",
      },
      {
        accessorKey: "name.lastName",
        header: "Last Name",
      },
      {
        accessorKey: "address", //normal accessorKey
        header: "Address",
      },
      {
        accessorKey: "city",
        header: "City",
      },
      {
        accessorKey: "state",
        header: "State",
      },
    ]);

    setData([
      {
        name: {
          firstName: "John",
          lastName: "Doe",
        },
        address: "261 Erdman Ford",
        city: "East Daphne",
        state: "Kentucky",
      },
      {
        name: {
          firstName: "Jane",
          lastName: "Doe",
        },
        address: "769 Dominic Grove",
        city: "Columbus",
        state: "Ohio",
      },
      {
        name: {
          firstName: "Joe",
          lastName: "Doe",
        },
        address: "566 Brakus Inlet",
        city: "South Linda",
        state: "West Virginia",
      },
      {
        name: {
          firstName: "Kevin",
          lastName: "Vandy",
        },
        address: "722 Emie Stream",
        city: "Lincoln",
        state: "Nebraska",
      },
      {
        name: {
          firstName: "Joshua",
          lastName: "Rolluffs",
        },
        address: "32188 Larkin Turnpike",
        city: "Charleston",
        state: "South Carolina",
      },
    ]);
  }, []);

  return <MaterialReactTable columns={columns} data={data} enableGrouping />;
};

export default Basic;

```

By using this simple code to reproduce the Show/Hide Columns function cannot be used.

### Minimal, Reproducible Example - (Optional, but Recommended)

use the above sample code I provided, the Show/Hide columns function shows empty.

### Terms

- [X] I understand that if my bug cannot be reliably reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by KevinVandy** [comment]
@realDNA I'm confused. What's the "malfunction"? I copied your code snippet into [this Stackblitz](https://stackblitz.com/edit/github-fz6nd1?file=src/TS.tsx) and it is working fine. All the columns can still be hidden

**Comment by KevinVandy** [comment]
@realDNA After looking a bit more, were you talking about the columns not showing up in the menu?

<img width="274" alt="image" src="https://user-images.githubusercontent.com/28243511/203875097-705019ca-0ca2-4295-a7e4-3dd830fe0786.png">

If so, that's an important screenshot to add when creating this issue. "Malfunction" is a very general term that can mean anything.

The cause is that you are not defining your columns for the first render. If you do it that way, you will need to manage the column order yourself like down below using the `state` prop.

```ts
return (
  <MaterialReactTable
    columns={columns}
    data={data}
    enableGrouping
    state={{ columnOrder: columns.map((c) => c.accessorKey) }}
  />
);
```

Otherwise, I don't recommend initializing state in a useEffect like that, especially for columns.

**Comment by realDNA** [comment]
Hi, Kevin Vandy,

Thanks for your help.

After adding state={{ columnOrder: columns.map((c) => c.accessorKey) }} to my code then it works!

Sorry, I tried hard to upload the screenshot, but it couldn't be uploaded. At first, I thought it was the GitHub issue, but this morning I realized it was somehow blocked by my firewall.

For the columns, because there are thousands of columns for users to choose from, I am not able to define them at the first rendering. It needs to be set after some processing.

Anyway, thanks for your help!
